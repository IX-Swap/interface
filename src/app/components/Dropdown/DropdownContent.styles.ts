import makeStyles from '@mui/styles/makeStyles'

export const useStyles = makeStyles(theme => ({
  popper: {
    zIndex: 1500,
    '&[data-popper-placement*="bottom"] $arrow': {
      top: 0,
      left: 0,
      marginTop: '-0.9em',
      width: '3em',
      height: '1em',
      '&::before': {
        borderWidth: '0 1em 1em 1em',
        borderColor: `transparent transparent ${theme.palette.background.paper} transparent`
      }
    },
    '&[data-popper-placement*="top"] $arrow': {
      bottom: 0,
      left: 0,
      marginBottom: '-0.9em',
      width: '3em',
      height: '1em',
      '&::before': {
        borderWidth: '1em 1em 0 1em',
        borderColor: `${theme.palette.background.paper} transparent transparent transparent`
      }
    },
    '&[data-popper-placement*="right"] $arrow': {
      left: 0,
      marginLeft: '-0.9em',
      height: '3em',
      width: '1em',
      '&::before': {
        borderWidth: '1em 1em 1em 0',
        borderColor: `transparent ${theme.palette.background.paper} transparent transparent`
      }
    },
    '&[data-popper-placement*="left"] $arrow': {
      right: 0,
      marginRight: '-0.9em',
      height: '3em',
      width: '1em',
      '&::before': {
        borderWidth: '1em 0 1em 1em',
        borderColor: `transparent transparent transparent ${theme.palette.background.paper}`
      }
    }
  },
  arrow: {
    position: 'absolute',
    fontSize: 7,
    width: '3em',
    height: '3em',
    '&::before': {
      content: '""',
      margin: 'auto',
      display: 'block',
      width: 0,
      height: 0,
      borderStyle: 'solid'
    }
  }
}))
