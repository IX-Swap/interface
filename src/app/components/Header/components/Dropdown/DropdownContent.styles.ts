import makeStyles from '@mui/styles/makeStyles'

export const useStyles = makeStyles(theme => ({
  popper: {
    zIndex: 1500,
    transform: 'translate(0, 80px)!important',
    borderTopLeftRadius: 0,
    borderTopRightRadius: 0,
    borderBottomLeftRadius: 8,
    borderBottomRightRadius: 8,
    '& > div': {
      borderTopLeftRadius: 0,
      borderTopRightRadius: 0,
      borderBottomLeftRadius: 8,
      borderBottomRightRadius: 8,
      border:
        theme.palette.mode === 'light'
          ? `1px solid ${theme.palette.secondary.light}`
          : 'none',
      boxShadow:
        theme.palette.mode === 'light'
          ? '0px 80px 80px rgba(162, 172, 191, 0.16)!important'
          : 'none!important'
    }
  }
}))
