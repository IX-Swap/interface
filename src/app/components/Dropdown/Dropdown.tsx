import React, { createElement, FunctionComponent } from 'react'
import PopupState, {
  bindPopper,
  bindTrigger,
  InjectedProps
} from 'material-ui-popup-state'
import { DropdownContent } from 'app/components/Dropdown/DropdownContent'
import { ThemeProvider } from '@material-ui/core/styles'
import { PopperPlacementType, Theme } from '@material-ui/core'
import { darkTheme } from 'themes/dark'
import { createAppTheme } from 'themes'
import { lightTheme } from 'themes/light'
import { useAppTheme } from 'hooks/useAppTheme'

export interface DropdownTriggerProps {
  triggerProps: ReturnType<typeof bindTrigger>
  injectedProps: InjectedProps
}

export interface DropdownContentProps {
  popperProps: ReturnType<typeof bindPopper>
  injectedProps: InjectedProps
}

export interface DropdownProps {
  trigger: FunctionComponent<DropdownTriggerProps>
  content: FunctionComponent<DropdownContentProps>
  arrow?: boolean
  placement?: PopperPlacementType
  contentTheme?: 'light' | 'dark'
}

export const Dropdown = (props: DropdownProps) => {
  const {
    content,
    trigger,
    arrow = false,
    placement = 'bottom-end',
    contentTheme
  } = props
  const defaultTheme = useAppTheme()
  const theme =
    contentTheme === undefined
      ? defaultTheme
      : createAppTheme(contentTheme === 'dark' ? darkTheme : lightTheme)

  return (
    <PopupState variant='popper'>
      {popupState => (
        <>
          {createElement(trigger, {
            triggerProps: bindTrigger(popupState),
            injectedProps: popupState
          })}

          <ThemeProvider theme={theme as Theme}>
            <DropdownContent
              popupState={popupState}
              placement={placement}
              arrow={arrow}
            >
              {content}
            </DropdownContent>
          </ThemeProvider>
        </>
      )}
    </PopupState>
  )
}
