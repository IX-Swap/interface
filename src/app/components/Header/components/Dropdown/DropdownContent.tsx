import React, { createElement } from 'react'
import { ClickAwayListener, Paper, Popper } from '@mui/material'
import { bindPopper, InjectedProps } from 'material-ui-popup-state'
import { useStyles } from 'app/components/Header/components/Dropdown/DropdownContent.styles'
import { DropdownProps } from 'app/components/Dropdown/Dropdown'

export interface DropdownContentProps {
  popupState: InjectedProps
  placement: DropdownProps['placement']
  children: DropdownProps['content']
}

export const DropdownContent = (props: DropdownContentProps) => {
  const { popupState, placement, children } = props
  const classes = useStyles({ placement: placement })
  const popperProps = bindPopper(popupState)
  const handleClickAway = () => {
    if (popupState.isOpen) {
      popupState.close()
    }
  }

  return (
    <Popper
      {...bindPopper(popupState)}
      placement={placement}
      className={classes.popper}
    >
      <ClickAwayListener onClickAway={handleClickAway}>
        <Paper>
          {createElement(children, {
            injectedProps: popupState,
            popperProps
          })}
        </Paper>
      </ClickAwayListener>
    </Popper>
  )
}
