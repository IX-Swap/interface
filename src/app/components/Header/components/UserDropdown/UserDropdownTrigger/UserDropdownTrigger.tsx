import React from 'react'
import { IconButton } from '@mui/material'
import { DropdownTriggerProps } from 'app/components/Dropdown/Dropdown'
import { CustomAvatar } from 'ui/CustomAvatar'
import { useStyles } from 'app/components/Header/components/UserDropdown/UserDropdownTrigger/UserDropdownTrigger.styles'
import classnames from 'classnames'
import { useAuth } from 'hooks/auth/useAuth'
import { useIndividualIdentity } from 'hooks/identity/useIndividualIdentity'

export const UserDropdownTrigger = (props: DropdownTriggerProps) => {
  const classes = useStyles({ isOpen: props.injectedProps.isOpen })
  const { user } = useAuth()
  const { data } = useIndividualIdentity()

  if (user === undefined) {
    return null
  }

  return (
    <IconButton
      {...props.triggerProps}
      color='inherit'
      aria-haspopup='true'
      aria-controls='profile-menu'
      size='large'
      disableTouchRipple
      className={classnames(classes.wrapper, {
        [classes.opened]: props.injectedProps.isOpen
      })}
    >
      <CustomAvatar documentId={data?.photo} ownerId={user._id}>
        {user.name[0]}
      </CustomAvatar>
    </IconButton>
  )
}
