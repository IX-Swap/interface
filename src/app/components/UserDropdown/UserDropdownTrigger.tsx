import React from 'react'
import UserIcon from 'assets/icons/navigation/user.svg'
import { IconButton } from '@material-ui/core'
import { DropdownTriggerProps } from 'app/components/Dropdown/Dropdown'

export const UserDropdownTrigger = (props: DropdownTriggerProps) => {
  return (
    <IconButton
      {...props.triggerProps}
      color='inherit'
      aria-haspopup='true'
      aria-controls='profile-menu'
    >
      <UserIcon />
    </IconButton>
  )
}
