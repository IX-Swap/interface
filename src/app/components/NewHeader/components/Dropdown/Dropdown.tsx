import React, { createElement, FunctionComponent } from 'react'
import PopupState, {
  bindPopper,
  bindToggle,
  InjectedProps
} from 'material-ui-popup-state'
import { PopperPlacementType } from '@mui/material'
import { DropdownContent } from 'app/components/NewHeader/components/Dropdown/DropdownContent'

export interface DropdownTriggerProps {
  triggerProps: ReturnType<typeof bindToggle>
  injectedProps: InjectedProps
}

export interface DropdownContentProps {
  popperProps: ReturnType<typeof bindPopper>
  injectedProps: InjectedProps
}

export interface DropdownProps {
  trigger: FunctionComponent<DropdownTriggerProps>
  content: FunctionComponent<DropdownContentProps>
  placement?: PopperPlacementType
  contentTheme?: 'light' | 'dark'
}

export const Dropdown = (props: DropdownProps) => {
  const { content, trigger, placement = 'bottom-end' } = props

  return (
    <PopupState variant='popper'>
      {popupState => (
        <>
          {createElement(trigger, {
            triggerProps: bindToggle(popupState),
            injectedProps: popupState
          })}
          <DropdownContent popupState={popupState} placement={placement}>
            {content}
          </DropdownContent>
        </>
      )}
    </PopupState>
  )
}
