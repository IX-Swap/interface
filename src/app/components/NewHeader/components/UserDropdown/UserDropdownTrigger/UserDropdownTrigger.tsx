import React from 'react'
import { IconButton } from '@mui/material'
import { DropdownTriggerProps } from 'app/components/Dropdown/Dropdown'
import { CustomAvatar } from 'ui/CustomAvatar'
import { useStyles } from 'app/components/NewHeader/components/UserDropdown/UserDropdownTrigger/UserDropdownTrigger.styles'
import classnames from 'classnames'
// TODO Remove mocked hook after demo
import { useAuth } from 'app/components/NewHeader/hooks/mock/useAuth'
// import { useAuth } from 'hooks/auth/useAuth'

export const UserDropdownTrigger = (props: DropdownTriggerProps) => {
  const classes = useStyles({ isOpen: props.injectedProps.isOpen })
  const { user } = useAuth()

  if (user === undefined) {
    return null
  }

  return (
    <IconButton
      {...props.triggerProps}
      color='inherit'
      aria-haspopup='true'
      aria-controls='profile-menu'
      size='large'
      disableTouchRipple
      className={classnames(classes.wrapper, {
        [classes.opened]: props.injectedProps.isOpen
      })}
    >
      {/* TODO Needs to add view photo functionality, as in old Avatar component */}
      <CustomAvatar>{user.name[0]}</CustomAvatar>
    </IconButton>
  )
}
