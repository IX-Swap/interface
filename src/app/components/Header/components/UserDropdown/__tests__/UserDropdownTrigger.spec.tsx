import React from 'react'
import { render } from 'test-utils'
import * as useAuth from 'hooks/auth/useAuth'
import * as useIndividualIdentity from 'hooks/identity/useIndividualIdentity'
import { individual } from '__fixtures__/identity'
import { generateQueryResult } from '__fixtures__/useQuery'
import { DropdownTriggerProps } from 'app/components/Header/components/Dropdown/Dropdown'
import { UserDropdownTrigger } from 'app/components/Header/components/UserDropdown/UserDropdownTrigger/UserDropdownTrigger'

jest.mock('@mui/material/Badge', () => jest.fn(() => null))

const successAuthObjResponse = {
  user: {
    _id: 'test',
    name: 'name'
  }
}

const failedAuthObjResponse = {
  user: undefined
}

describe('UserDropdownTrigger', () => {
  const props: DropdownTriggerProps = {
    injectedProps: { close: jest.fn() },
    triggerProps: {}
  } as any

  afterEach(async () => {
    jest.clearAllMocks()
  })

  it('renders empty component if user is undefined', () => {
    jest
      .spyOn(useAuth, 'useAuth')
      .mockImplementation(() => failedAuthObjResponse as any)
    const { container } = render(<UserDropdownTrigger {...props} />)

    expect(container).toBeEmptyDOMElement()
  })

  it('should match snapshot if user is not undefined and has identity info', () => {
    const useIndividualIdentityResponse = generateQueryResult({
      data: individual,
      isLoading: true
    })

    jest
      .spyOn(useAuth, 'useAuth')
      .mockImplementation(() => successAuthObjResponse as any)
    jest
      .spyOn(useIndividualIdentity, 'useIndividualIdentity')
      .mockImplementation(() => useIndividualIdentityResponse as any)

    const { container } = render(<UserDropdownTrigger {...props} />)
    expect(container).toMatchSnapshot()
  })
})
