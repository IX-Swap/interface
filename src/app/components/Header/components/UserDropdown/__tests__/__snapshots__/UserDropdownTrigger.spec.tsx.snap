// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDropdownTrigger should match snapshot if user is not undefined and has identity info 1`] = `
<div>
  <button
    aria-controls="profile-menu"
    aria-haspopup="true"
    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge makeStyles-wrapper-3 css-1ncy747-MuiButtonBase-root-MuiIconButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-69x3yt-MuiAvatar-root"
      style="width: 32px; height: 32px; max-width: initial;"
    >
      n
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
